### ISSUE TITLE
 How to use the Twitter Premium API with Tweetinvi?

### ISSUE THREAD

**Issue by Pumation** [linvi/tweetinvi#689]
I have used Tweetinvi before for the free Twitter API and it has worked perfectly for me, but now I have registered to use the Premium API and get more Tweets of up to 30 days before, the situation is that I have not found any way to use Tweetinvi With this Twitter API, Someone knows some way of adapting or some other library or API that works with the Premium version (Sandbox).

Greetings.

**Comment by linvi** [comment]
Hello,

Tweetinvi does not offer access to the API as I have personally never been able to obtain access to such an account.

If you are happy to work with me and provide access to allow me to test and improve the API, I will be happy to help.

Cheers,
Linvi

**Comment by PunzunLtd** [comment]
At least with the Search API, the line to refactor to get the 30-day and unlimited APIs is in 'SearchQueryGenerator :: GetSearchTweetsQuery` at 39. The constant value `Resources.Search_SearchTerms` needs to be a mutable option. For the 30-day API, the URL ends with `search/30day/:label.json` and for unlimited it's `search/fullarchive/:label.json`, where `:label.json` is either your environment or is left off for production. (See [Twitter Premium docs](https://developer.twitter.com/en/docs/tweets/search/api-reference/premium-search).

In other words, the URL needs to be configurable, not constant. It would also work if we could inject `Resources` as a custom file—but configurable seems to make more sense to me, as it could vary by environment.

**Comment by linvi** [comment]
@PunzunLtd 

Hello, 

Thank you for reaching out. Tweetinvi 5.0 will offer better control over the requests being sent via client events. But also with the ability to send simple requests (not yet documented). 

Unfortunately I have still not received any access for testing premium features so I cannot provide such feature before I do (or anyone offer me to access their credentials). Feel free to support my request here : https://twittercommunity.com/t/enterprise-account-ands-ads-for-open-source-developers/136289/2

In the meantime Tweetinvi 5.0 will allow you to do something close to what you are looking for.

[Documentation for direct calls](https://linvi.github.io/tweetinvi/dist/features/direct-api-calls.html)

``` c#
var result = await client.Execute.Request<TweetDTO[]>(request =>
{
    request.Url = "https://api.twitter.com/1.1/statuses/home_timeline.json";
    request.HttpMethod = HttpMethod.GET;
});

// TweetDTO[]
var tweetDtos = result.DataTransferObject;

// ITweet[]
var tweets = client.Factories.CreateTweets(tweetDtos);
```

**Comment by PunzunLtd** [comment]
I think the general fix is to make the URLs configurable. This same fix would work for using Developer accounts, which you should probably have. See [premium search documentation](https://developer.twitter.com/en/docs/tweets/search/api-reference/premium-search) which also demonstrates how to search developer account endpoints as well.

When do you expect to release this feature? Is it already in Git, so I can temporarily build local DLLs until the release?

**Comment by linvi** [comment]
Yes, this is available in master branch.

**Comment by PunzunLtd** [comment]
@linvi , would it be possible to publish a snapshot of the current v5 bits in a pre-release NuGet package?

**Comment by linvi** [comment]
@PunzunLtd I am currently fixing an issue with iterators over the search engine. I will try and deploy a nuget pre-release tonight or tomorrow.

Cheers,
Linvi

**Comment by linvi** [comment]
@PunzunLtd Tweetinvi 5.0 is now available in alpha package.

https://www.nuget.org/packages/TweetinviAPI/5.0.0-alpha-1

**Comment by PunzunLtd** [comment]
Awesome! Thanks!

**Comment by linvi** [comment]
@PunzunLtd After you have played with the new version of the library, If you could provide some feedback I will be very grateful.

Cheers,
Linvi

**Comment by PunzunLtd** [comment]
There's another wrinkle. Enterprise requests require Basic authentication in the GET or POST authorization headers: https://developer.twitter.com/en/docs/tweets/search/api-reference/enterprise-search.

**Comment by linvi** [comment]
@PunzunLtd

Hello there,

Have you tried adding the header to the request.

``` c#
var username = "Your username";
var password = "Your password";
var basicAuthContent = Convert.ToBase64String(ASCIIEncoding.ASCII.GetBytes(username + ":" + password));
var result = await client.Execute.Request(request =>
{
    request.Url = "https://....";
    request.CustomHeaders.Add("Authorization", "Basic " + basicAuthContent);
});
```

**Comment by linvi** [comment]
@PunzunLtd The code above won't work as `Authorization` would be overriden.
I will release a bug fix for this.

Ticket : https://github.com/linvi/tweetinvi/issues/1037

**Comment by linvi** [comment]
@PunzunLtd alpha-2 has been released. Note that names of async methods have a new `Async` postfix as recommend on gitter (#1038).

You should now be able to customize the authentication header.

**Comment by tinsoldier** [comment]
It more or less works in 5.0.3, you end up having to provide a custom wrapper for the results as the API returns a JSON object with a "results" array property with the actual tweetDTOs versus the "statuses" array the standard search API uses.

Incidentally, I'm unsure if there's a better way to provide the search parameters, I'm just using a UriBuilder to construct the query string.

Are you still looking at testing the enterprise apis?

```C#
// This works fine
var client = new TwitterClient(null); // :)
var url = "https://gnip-api.twitter.com/search/30day/accounts/accountname/prod.json?query=yourquery";
var username = "username";
var password = "password";
var basicAuthContent = Convert.ToBase64String(ASCIIEncoding.ASCII.GetBytes($"{username}:{password}"));
var result = await client.Execute.RequestAsync<TwitterResult>(request =>
{
	request.Url = url;
	request.HttpMethod = HttpMethod.GET;
	request.CustomHeaders.Add("Authorization", "Basic " + basicAuthContent);
});

// Custom TwitterResult wrapper
var tweetDtos = result.Model;

// ITweet[]
var tweets = client.Factories.CreateTweets(tweetDtos.results);

public class TwitterResult
{
	public List<TweetDTO> results;
	public string next;
}
```

**Comment by linvi** [comment]
> Are you still looking at testing the enterprise apis?

Thanks for your comments and yes, I am still looking to provide support for the Premium API. If you can help in that regard it would be great.

I am also keen on having Pull Requests to offer the features needed for Premium.
